### PR TITLE
Change return signature of router navigate() and navigateToRoute() from boolean to NavigationResult

### PR DESCRIPTION
**Pull request by matthewcorven**
Fix for #648 to correct typings.  Calllers of navigate()/navigateToRoute() could await the routing to complete, but didn't know it due to typings.
@davismj

### CODE REVIEW

**Comment by matthewcorven** [comment]
@davismj and perhaps type `NavigationResult` would be more clearly named a `AwaitableNavigationResult`?

File: test/router.spec.ts [hunk 1]
```diff
@@ -230,6 +232,40 @@ describe('the router', () => {
           done();
         });
     });
+
+    it('should return to caller immediately if not awaited', (done) => {
```
Comment (davismj) [inline test/router.spec.ts]: well, these tests don't really test anything about aurelia. only how async works. they will never fail.

File: src/router.ts [hunk 2]
```diff
@@ -310,7 +310,7 @@ export class Router {
    * @param params The route parameters to be used when populating the route pattern.
    * @param options The navigation options.
    */
-  navigateToRoute(route: string, params?: any, options?: NavigationOptions): boolean {
+  navigateToRoute(route: string, params?: any, options?: NavigationOptions): NavigationResult {
```
Comment (davismj) [inline src/router.ts]: Strangely in my local version, this is what I already had. I wonder when and why they got changed.

File: test/router.spec.ts (hunk 1 above)
Comment (matthewcorven) [reply test/router.spec.ts]: Good call, removed

**Comment by davismj** [comment]
@matthewcorven thx for the PR, definitely helps me see what's up.

I think the fix needs to be a bit more robust here. It starts in history / browser where navigate has a boolean return type (https://github.com/aurelia/history/blob/master/src/index.js#L56, https://github.com/aurelia/history-browser/blob/master/src/browser-history.ts#L54), which is only accepted because of other insufficient typings around history options (https://github.com/aurelia/history-browser/blob/master/src/browser-history.ts#L111, https://github.com/aurelia/history-browser/blob/master/src/browser-history.ts#L297, https://github.com/aurelia/history-browser/blob/master/src/browser-history.ts#L54).

Additionally, I don't like the `NavigationResult` typing. Consuming one requires testing for a `Promise` first: `const result: boolean = await Promise.resolve(this.router.navigate('foo');`. This implementation detail should not be the burden of an end developer, who should simply be able to write `const result: boolean = await this.router.navigate('foo');`.

I'll go ahead and fix this in all three libraries by:
- Adding a typing for `NavigationResult` in aurelia-history
- Adding a typing for `HistoryOptions` in aurelia-history-browser
- Enforcing `Promise<NavigationResult>` return types on all navigation methods in all libraries

This will be a breaking change. Currently, the `NavigationResult` type includes `PipelineResult | boolean`, which is a difficult type to work with. I will unify them as `NavigationResult = { completed: boolean }`, and transform all flat `boolean` returns into a `NavigationResult`. This means, moving forward, any navigation results, no matter where they come from, can be consumed like this:

`const { completed } = await router.navigate('foo');`

**Comment by matthewcorven** [comment]
Thanks @davismj !

**Pull request closed without merging**